File: cellviz/src/cellviz/plots/voltage_time.py
from __future__ import annotations

from typing import Any

import plotly.graph_objects as go

from ..io import CellData, CyclerName, load_many
from ..style import cell_color


def _cycle_voltage_time(cell: CellData, cycles: list[int] | None, max_cycles: int):
    out: list[tuple[int, list[float], list[float]]] = []
    allowed = set(cycles) if cycles else None
    cycle_keys = sorted(cell.curves.keys())[:max_cycles]
    for cyc in cycle_keys:
        if allowed is not None and cyc not in allowed:
            continue
        curve = cell.curves[cyc]
        v = curve.get("Voltage [V]") or curve.get("voltage")
        t = curve.get("Time [s]") or curve.get("time")
        if v is None:
            continue
        times: list[float] = []
        volts: list[float] = []
        for j, vv in enumerate(v):
            if vv is None:
                continue
            tt = t[j] if (t and j < len(t) and t[j] is not None) else j
            times.append(float(tt))
            volts.append(float(vv))
        if len(times) >= 2:
            out.append((cyc, times, volts))
    return out


def voltage_time_plot(
    data: Any,
    *,
    cycles: list[int] | None = None,
    max_cycles: int = 5,
    labels: list[str] | None = None,
    cycler: CyclerName = "auto",
) -> go.Figure:
    """Plot voltage vs time for the first few cycles per cell."""
    cells = load_many(data, labels=labels, cycler=cycler)
    multi = len(cells) > 1
    traces: list[go.Scatter] = []
    for i, cell in enumerate(cells):
        rows = _cycle_voltage_time(cell, cycles, max_cycles)
        if not rows:
            continue
        base = cell.color or cell_color(cell.cell_id or cell.label or f"dataset-{i+1}", i)
        prefix = f"{cell.label} — " if multi else ""
        for cyc, times, volts in rows:
            traces.append(
                go.Scatter(
                    x=times,
                    y=volts,
                    mode="lines",
                    name=f"{prefix}Cycle {cyc}",
                    line={"width": 1.5, "color": base},
                    hovertemplate=(
                        f"{cell.label} Cycle {cyc}<br>Time: %{{x:.1f}} s<br>"
                        "Voltage: %{y:.3f} V<extra></extra>"
                    ),
                )
            )

    fig = go.Figure(data=traces)
    fig.update_layout(
        xaxis={"title": {"text": "Time (s)"}},
        yaxis={"title": {"text": "Voltage (V)"}},
        showlegend=True,
    )
    return fig

File: packages/cellseer-lib/src/plotting/voltageTime.ts
import type { BuildVoltageTimeOpts, RecordDataset, VoltageTimeFigure } from '../types';

const DEFAULT_MAX_CYCLES = 5;

export function buildVoltageTimeFigure(
  datasets: RecordDataset[],
  opts: BuildVoltageTimeOpts = {},
): VoltageTimeFigure {
  const maxCycles = opts.maxCycles ?? DEFAULT_MAX_CYCLES;
  const traces: Plotly.Data[] = [];

  datasets.forEach((dataset) => {
    const curveKeys = Object.keys(dataset.curves).slice(0, 200);
    const cycles = curveKeys
      .map((k) => parseInt(k, 10))
      .filter((x) => !isNaN(x))
      .sort((a, b) => a - b)
      .slice(0, maxCycles);
    const namePrefix = datasets.length > 1 ? `${dataset.label} — ` : '';
    const fallbackColor = dataset.color ?? '#6b7280';

    let firstCycleRendered = false;
    for (const cyc of cycles) {
      const curve = dataset.curves[String(cyc)];
      if (!curve) continue;
      const vKey = 'Voltage [V]' in curve ? 'Voltage [V]' : 'voltage';
      const tKey = 'Time [s]' in curve ? 'Time [s]' : null;
      const v = (curve[vKey] ?? []) as (number | null)[];
      const tRaw = tKey ? ((curve[tKey] ?? []) as (number | null)[]) : null;
      const n = v.length;
      const times: number[] = [];
      const volts: number[] = [];
      for (let j = 0; j < n; j++) {
        const vv = v[j];
        if (vv == null) continue;
        const tt = tRaw && tRaw[j] != null ? (tRaw[j] as number) : j;
        times.push(tt);
        volts.push(vv as number);
      }
      if (times.length >= 2) {
        const isFirst = !firstCycleRendered;
        firstCycleRendered = true;
        traces.push({
          x: times,
          y: volts,
          type: 'scatter',
          mode: 'lines',
          name: isFirst ? dataset.label : `${namePrefix}Cycle ${cyc}`,
          showlegend: isFirst,
          line: { width: 1.5, color: fallbackColor },
          hovertemplate: `${dataset.label} Cycle ${cyc}<br>Time: %{x:.1f} s<br>Voltage: %{y:.3f} V<extra></extra>`,
        });
      }
    }
  });

  return {
    data: traces,
    layout: {},
  };
}
